apps/extension/src/storage/base.ts

```diff
@@ -69,18 +94,47 @@ export class ExtensionStorage<T> {
   }
 
   private async migrateIfNeeded<K extends keyof T>(key: K, item: StorageItem<T[K]>): Promise<T[K]> {
-    if (item.version !== this.version) {
-      const migrationFn = this.migrations[key]?.[item.version];
-      if (migrationFn) {
-        // Update the value to latest schema
-        const transformedVal = migrationFn(item.value) as T[K];
-        await this.set(key, transformedVal);
-        return transformedVal;
-      } else {
-        // If there's no migration function, handle it appropriately, possibly by just returning the current value
-        return item.value;
-      }
+    // The same version means no migrations are necessary
+    if (item.version === this.version) {
+      return item.value;
     }
-    return item.value;
+
+    // Check for an ongoing migration, if exists, return the result of that
+    const ongoingMigration = this.migrationLocks[String(key)] as Promise<T[K]> | undefined;
+    if (ongoingMigration) {
+      return ongoingMigration;
+    }
+
+    // Store the promise in the lock map. Ensures multiple callers don't run migrations multiple times.
+    const migrationPromise = this.migrateField(key, item);
+    this.migrationLocks[String(key)] = migrationPromise;
```

Comment:
New addition. Given multiple parts of the codebase are accessing fields at the same time that require migration, we need a way to ensure multiple migrations are not run and thus share the same promise to be resolved.

Reply:
locking mechanism implementation changed, but rationale remains